### PR TITLE
Fix pickling

### PR DESCRIPTION
**Pull request by mehrdadn**
This pull request is intended to fix problems that we have encountered with the existing pickling frameworks:

	def f():
		def g(): return g
		return g
	pickling.loads(pickling.dumps(f()))  # fails, but should succeed

	import typing
	pickling.loads(pickling.dumps(typing.List[int]))  # fails, but should succeed

### CODE REVIEW

**Comment by robertnishihara** [comment]
If I `cd lib/python/ray` and in `python` I do `import pickling`, I get the following error.
```python
In [1]: import pickling
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-1-0b2666f80c64> in <module>()
----> 1 import pickling

/Users/rkn/Workspace/ray/lib/python/ray/pickling.py in <module>()
      5 from cloudpickle import pickle, cloudpickle, CloudPickler, load, loads
      6 
----> 7 python_dll = cdll.python27
      8 
      9 def dump(obj, file, protocol=2):

/Users/rkn/anaconda/lib/python2.7/ctypes/__init__.pyc in __getattr__(self, name)
    433         if name[0] == '_':
    434             raise AttributeError(name)
--> 435         dll = self._dlltype(name)
    436         setattr(self, name, dll)
    437         return dll

/Users/rkn/anaconda/lib/python2.7/ctypes/__init__.pyc in __init__(self, name, mode, handle, use_errno, use_last_error)
    363 
    364         if handle is None:
--> 365             self._handle = _dlopen(self._name, mode)
    366         else:
    367             self._handle = handle

OSError: dlopen(python27, 6): image not found
```

**Comment by robertnishihara** [comment]
After the most recent changes, I can successfully import `pickling` :) If I do (in `python scripts/shell.py`)
```python
In [1]: @ray.remote([int], [int])
   ...: def f(n):
   ...:     return f(n)
   ...: 

In [2]: ray.get(f(1))
Exception: The task that created this object reference failed with error message:
          Traceback (most recent call last):
            File "<ipython-input-1-00baffb3675a>", line 3, in f
          NameError: global name 'f' is not defined
```

**Comment by mehrdadn** [comment]
Did this use to work? If not, can we make it another PR?

**Comment by robertnishihara** [comment]
Oh sure, I thought the point of this PR was to make that work.

**Comment by mehrdadn** [comment]
Nah this was to make self references in an inner scope work, not global scope.

**Comment by mehrdadn** [comment]
It appears that you've opened a huge can of worms. =)

Explanation is below. Scroll further to see the possible solution(s).

~~~~~~~~~~~~~~~~~~~~~

The immediate issue is that the function is **not** defined until *after* the function has been decorated.
Hence, the `globals` that should include the function does not include it when it is pickled.

Now, this is easy to make this case work (and for that, skip to the bottom), but the bigger problem is much more subtle and more difficult to work around conceptually. To see why, consider this example:

	@ray.remote([int], [int])
	def f(x): return g(x)
	
	@ray.remote([int], [int])
	def g(x): return x

	ray.get(f(2))

We'd like this to work, but it wouldn't normally. Why? Because we send remote functions one-by-one, and after unpickling, the `__globals__` of every function becomes an independent dictionary, and this dictionary will never be updated again. Therefore, what we construct is **not** the same `f` and `g` as we have on the driver -- those actually *share* the same `__globals__` dictionary and are updated every time we define a new function. Once they are unpickled, however, their relationship to each other is lost.

The obvious way to make this work is to set the `__globals__` of each function during the *unpickling* process, and update it as each function is sent over. However, *it is not obvious to me that this is the right solution*. Consider this code:

	h = [lambda x: x]

	@ray.remote([int], [int])
	def f(x): return g(h[0](x))

	h[0] = lambda x: 0

	@ray.remote([int], [int])
	def g(x): return x

	ray.get(f(2))

What exactly is the correct behavior here? Do we need to *redefine* `h` after unpickling? Or does that mean we should automagically somehow *modify* the corresponding `h`'s elements when `g` is defined remotely? Or do we treat `f` as a black box and not touch its `h` at all, leaving it the way `f` saw it when it was defined? And in the last case, what if `h` wasn't defined until later? Should that fail?

~~~~~~~~~~~~~~~~~~~~~

Long story short, the most consistent solutions I can think of for now are:

1. Disallow referring to any global variable that has not yet been defined.

2.  Allow the function to refer to itself as a global, but disallow anything defined afterward.

I took approach #2 (since it is a strict superset of #1's behaviors) and updated the code.
Here's a sample test -- make sure to start 3 workers to see it work:

	@ray.remote([int], [int])
	def fact(x): return x * ray.get(fact(x - 1)) if x > 1 else x

	ray.get(fact(3))  # 6

**Comment by robertnishihara** [comment]
Did you mean to say 
```python
ray.get(fact2(2))  # = 2
ray.get(fact2(3))  # fails
```

**Comment by mehrdadn** [comment]
No? That should be fine.

**Comment by mehrdadn** [comment]
Ah right, I think I see what's happening now, though I'm not sure how it worked the last time I tried it.
The thing is when `fact2` is called, the `fact1` it calls subsequently is the *decorated* `fact1`, whose only job is `_submit_task`... at which point the scheduler calls the *original* `fact1` (which was the one registered -- not the new one!) and that one fails to find `fact2` in its own `__globals__`.

I would say let's leave this as-is for now and decide how to handle such cases later, and tell users that all remote functions called must be (recursively) fully defined at the point of call.
